[CSS] :scope matching for multiple scoping roots

mdubet opened the pull request:
#### 0bc5d3ab92365420fc03f522bc29a6e8c6f9025c
<pre>
[CSS] :scope matching for multiple scoping roots
<a href="https://bugs.webkit.org/show_bug.cgi?id=265288">https://bugs.webkit.org/show_bug.cgi?id=265288</a>
<a href="https://rdar.apple.com/118748899">rdar://118748899</a>

Reviewed by Antti Koivisto.

<a href="https://drafts.csswg.org/css-cascade-6/#scope-limits">https://drafts.csswg.org/css-cascade-6/#scope-limits</a>

This patch generates multiple matched rules, one for each scoping root.
They will all participate in the selector matching with a distinct :scope, and will be selected
based on specificity (and proximity in a future patch).

* LayoutTests/TestExpectations:
* LayoutTests/fast/css/scope-at-rule-expected.html:
* LayoutTests/fast/css/scope-at-rule.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-container-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-evaluation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-focus-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-hover-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-implicit-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-media-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-nesting-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-supports-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-visited-cssom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/conditional-rules.html:
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType):
(WebCore::SelectorCompiler::SelectorCodeGenerator::generateElementIsScopeRoot):

Remove the assumption that :scope always means :root outside of SelectorQuery

* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::ruleMatches):
(WebCore::Style::ElementRuleCollector::collectMatchingRulesForList):
(WebCore::Style::ElementRuleCollector::scopeRulesMatch):
* Source/WebCore/style/ElementRuleCollector.h:

Canonical link: <a href="https://commits.webkit.org/271508@main">https://commits.webkit.org/271508@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2f604e3fa6b6b0a76eb86eb5bd703d822e800231

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28589 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7234 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29972 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/31237 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/26113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29093 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9347 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4604 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/31237 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28859 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5976 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24583 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/31237 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5348 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25584 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/32567 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26176 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26029 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/31606 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5316 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3483 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/29383 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6921 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/25378 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6852 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5777 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5835 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/ElementRuleCollector.cpp:
```diff
@@ -551,9 +557,17 @@ void ElementRuleCollector::collectMatchingRulesForList(const RuleSet::RuleDataVe
+        if (scopingRoots && !scopingRoots->isEmpty()) {
+            for (const auto* scopingRoot : *scopingRoots)
+                addRuleIfMatches(scopingRoot);
+        } else
```
maybe just `continue` instead of else.

anttijk commented on Source/WebCore/style/ElementRuleCollector.cpp:
```diff
@@ -551,9 +557,17 @@ void ElementRuleCollector::collectMatchingRulesForList(const RuleSet::RuleDataVe
+                addMatchedRule(ruleData, specificity, matchRequest);
+        };
+
+        if (scopingRoots && !scopingRoots->isEmpty()) {
```
scopeRulesMatch  could just return nullopt in the empty case

anttijk commented on Source/WebCore/style/ElementRuleCollector.cpp:
```diff
@@ -582,61 +596,91 @@ bool ElementRuleCollector::containerQueriesMatch(const RuleData& ruleData, const
 }
 
-bool ElementRuleCollector::scopeRulesMatch(const RuleData& ruleData, const MatchRequest& matchRequest)
+std::pair<bool, Vector<const Element*>>  ElementRuleCollector::scopeRulesMatch(const RuleData& ruleData, const MatchRequest& matchRequest)
```
maybe this stuff could be factored into ScopeEvaluator or similar in the future.

anttijk commented on Source/WebCore/style/ElementRuleCollector.h:
```diff
@@ -118,9 +118,9 @@ class ElementRuleCollector {
     void collectMatchingRules(const MatchRequest&);
     void collectMatchingRulesForList(const RuleSet::RuleDataVector*, const MatchRequest&);
-    bool ruleMatches(const RuleData&, unsigned& specificity, ScopeOrdinal);
+    bool ruleMatches(const RuleData&, unsigned& specificity, ScopeOrdinal, const Element* scopingRoots = nullptr);
```
scopingRoot

anttijk commented on Source/WebCore/style/ElementRuleCollector.cpp:
```diff
@@ -471,7 +471,7 @@ inline bool ElementRuleCollector::ruleMatches(const RuleData& ruleData, unsigned
         SelectorCompiler::compileSelector(compiledSelector, ruleData.selector(), SelectorCompiler::SelectorContext::RuleCollector);
 
-    if (compiledSelector.status == SelectorCompilationStatus::SimpleSelectorChecker) {
+    if (compiledSelector.status == SelectorCompilationStatus::SimpleSelectorChecker && !scopingRoot) {
```
Is this really needed? If the selector contains :scope then the status should be SelectorCheckerWithCheckingContext. If it does not then SimpleSelectorChecker should work fine without the scope too.

In any case this should be reflected in SelectorCompilationStatus and not require extra checking.

mdubet replied:
The SelectorCompiler was assuming that `:scope` is always `:root` outside of Selector Query

mdubet committed:
[CSS] :scope matching for multiple scoping roots
https://bugs.webkit.org/show_bug.cgi?id=265288
rdar://118748899

Reviewed by Antti Koivisto.

https://drafts.csswg.org/css-cascade-6/#scope-limits

This patch generates multiple matched rules, one for each scoping root.
They will all participate in the selector matching with a distinct :scope, and will be selected
based on specificity (and proximity in a future patch).

* LayoutTests/TestExpectations:
* LayoutTests/fast/css/scope-at-rule-expected.html:
* LayoutTests/fast/css/scope-at-rule.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-container-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-evaluation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-focus-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-hover-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-implicit-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-media-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-nesting-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-supports-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-visited-cssom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/conditional-rules.html:
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType):
(WebCore::SelectorCompiler::SelectorCodeGenerator::generateElementIsScopeRoot):

Remove the assumption that :scope always means :root outside of SelectorQuery

* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::ruleMatches):
(WebCore::Style::ElementRuleCollector::collectMatchingRulesForList):
(WebCore::Style::ElementRuleCollector::scopeRulesMatch):
* Source/WebCore/style/ElementRuleCollector.h:

Canonical link: https://commits.webkit.org/271508@main
